src/c/src/utf16.c

```diff
@@ -0,0 +1,161 @@
+// Copyright 2017 The Procyon Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "utf8.h"
+#include "utf16.h"
+#include "common.h"
+
+#include <string.h>
+
+#define UFFFD_REPLACEMENT_CHARACTER 0xfffd;
+
+static bool is_high_surrogate(uint16_t code_point) { return code_point >= 0xd800 && code_point <= 0xdbff; }
+static bool is_low_surrogate(uint16_t code_point) { return code_point >= 0xdc00 && code_point <= 0xdfff; }
+
+int32_t pn_rune16(const uint16_t* data, size_t size, size_t index) {
+    data += index;
+    size -= index;
+
+    if (is_low_surrogate(data[0])) {
+        return UFFFD_REPLACEMENT_CHARACTER;
+    }
+    if (!is_high_surrogate(data[0])) {
+        return data[0];
+    }
+    if (size < 2 || !is_low_surrogate(data[1])) {
+        return UFFFD_REPLACEMENT_CHARACTER;
+    }
+    pn_rune_t rune = (data[0] << 10) + data[1] - ((0xd800 << 10) + 0xdc00 - 0x10000);
+    return rune;
+}
+
+size_t pn_rune16_next(const uint16_t* data, size_t size, size_t index) {
+    data += index;
+    size -= index;
+
+    if (is_low_surrogate(data[index])) {
+        return index + 1;
+    }
+    if (!is_high_surrogate(data[0])) {
+        return index + 1;
+    }
+    if (size < 2 || !is_low_surrogate(data[1])) {
+        return index + 1;
+    }
+    return index + 2;
```

Comment:
This could be much simpler than the UTF-8 version:
```suggestion
    if (size >= 2 && is_high_surrogate(data[0]) && is_low_surrogate(data[1])) {
        return index + 2;
    }
    return index + 1;
```